juju/juju
# Upgrade Series Pins Leadership

## Description of change

This patch builds on previous leadership pinning work to expose pinning via a client API and recruit it in the upgrade-series command in order to pin applications represented by units running on the target machine.

When pinning we can derive the list of applications represented on the machine by using the list of units returned by the validation step.

For unpinning, a new method, _Applications_ is added to the _machinemanager_ API to get the target list.

Failures during prepare result in calls to unpin leadership for the machine's applications, but this patch does not expressly mitigate a failure scenario that occurs _in the middle_ of unpinning leadership. That will be discussed and implemented in future work.

The new client-side API for leadership added here, is the first example of using a structure that mirrors the API server side. Hence its residence in _/api/facades/client/leadership/_.

## QA steps

- `export JUJU_DEV_FEATURE_FLAGS=upgrade-series`.
- Bootstrap to LXD
- `juju deploy redis`
- `juju add-unit redis`
- `juju upgrade-series prepare 0`
- Check that the console output indicates frozen leadership for "redis".
- Use lxc to manually stop machine 0.
- Wait and check that leadership does not change from one Redis unit to another.
- Restart the machine with lxc.
- `juju upgrade-series complete 0`
- Check that the console output indicates unfrozen leadership for "redis".
- Manually stop the machine again.
- Wait to check that leadership switches to the unit on the running machine.

## Documentation changes

None.

## Bug reference

N/A
